devdoc/thandle_requirements.md

```diff
@@ -0,0 +1,135 @@
+
+Given a type `LL` previously introduced by `MU_DEFINE_STRUCT`, `thandle`'s macros will encapsulate the type, provide a HANDLE (that is, opaque type) to it which other APIs can use. The handle type is `THANDLE(LL)`.
+
+`thandle` will provide macros for declaring all the needed constructs in a .h file and macros for defining the constructs in a .c file.
+
+## Exposed API
+
+```c
+
+/*to be used as the type of handle that wraps T*/
+#define THANDLE(T)
+
+/*to be used in a header file*/
+#define THANDLE_TYPE_DECLARE(T)
+
+/*to be used in a .c file*/
+#define THANDLE_TYPE_DEFINE(T)
+
+```
+
+## THANDLE(T)
+
+```c
+#define THANDLE(T) 
+```
+
+`THANDLE(T)` introduces a new incomplete type that can be used with all subsequent APIs constructed with `THANDLE`. `T` is a type that was introduced by `MU_DEFINE_STRUCT` in a .c file and was wrapped using `THANDLE_TYPE_DEFINE(T)` in the .c file.
+
+The type is const - so the following incorrect assignment fails at compile time:
+
+```c
+    THANDLE(T) one = t_create(...);
+    THANDLE(T) two;
+    two = one; /*compiler error*/
+```
+
+## THANDLE_TYPE_DECLARE(T)
+```c
+#define THANDLE_TYPE_DECLARE(T)
+```
+
+`THANDLE_TYPE_DECLARE` introduces several functions that can be used with `THANDLE(T)` type. These are `THANDLE_DEC_REF(T)`, `THANDLE_INC_REF(T)`, `THANDLE_ASSIGN(T)`, `THANDLE_INITIALIZE(T)`.
+
+###  THANDLE_DEC_REF(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_DEC_REF(T), THANDLE(T), t);
+```
+
+`THANDLE_DEC_REF` decrements the reference count of `t` and frees memory if reference count reaches 0.
+
+**SRS_THANDLE_02_001: [** If `t` is `NULL` then `THANDLE_DEC_REF` shall return. **]**
+
+**SRS_THANDLE_02_002: [** `THANDLE_DEC_REF` shall decrement the ref count of `t`.  **]**
+
+**SRS_THANDLE_02_003: [** If the ref count of `t` reaches 0 then `THANDLE_DEC_REF` shall call `dispose` (if not `NULL`) and free the used memory.  **]**
+
+### THANDLE_INC_REF(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_INC_REF(T), THANDLE(T), t);
+```
+
+`THANDLE_INC_REF` increments the reference count of `t`.
+
+**SRS_THANDLE_02_004: [** If `t` is `NULL` then `THANDLE_INC_REF` shall return. **]**
+
+**SRS_THANDLE_02_005: [** `THANDLE_INC_REF` shall increment the reference count of `t`. **]**
+
+### THANDLE_ASSIGN(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_ASSIGN(T), THANDLE(T) *, t1, THANDLE(T), t2 );
+```
+
+`THANDLE_ASSIGN` does t1=t2. Both `t1` and `t2` are existing constructed handles.
+
+
+**SRS_THANDLE_02_006: [** If `t1` is `NULL` then `THANDLE_ASSIGN` shall return. **]**
+
+**SRS_THANDLE_02_007: [** If `*t1` is `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall return. **]**
+
+**SRS_THANDLE_02_008: [** If `*t1` is `NULL` and `t2` is not `NULL` then `THANDLE_ASSIGN` shall increment the reference count of `t2` and store `t2` in `*t1`. **]**
+
+**SRS_THANDLE_02_009: [** If `*t1` is not `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall decrement the reference count of `*t1` and store `NULL` in `*t1`. **]**
```

Comment:
>If `*t1` is not `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall decrement the reference count of `*t1` and store  [](start = 26, length = 119)

I find this very interesting. And nice. And at the same time I am a bit scared of it as these assignment will hide another layer of inc ref/dec ref. Let's see how it goes.

Reply:
Frankly, initially I did not even have inc_REF, because "you never want to inc-ref just because". But then I thought "so what happens when the hadle is passed as context?"

So I needed INC_REF. :(

And yes, I agree, let's see how it goes.

---
In reply to: [385367955](https://github.com/Azure/azure-c-util/pull/16#discussion_r385367955) [](ancestors = 385367955)